File: backend/src/controllers/task.controller.ts
import { Request, Response } from "express";
import { TaskService } from "../services/task.service";
import { AuthenticatedRequest } from "../middlewares/auth.middleware";
import { TaskStatus } from "../utils/constants";

export class TaskController {
  static async createTask(req: AuthenticatedRequest, res: Response) {
    try {
      const {
        title,
        description,
        teamId,
        moduleId,
        plannedStartDate,
        plannedEndDate,
        clientId,
        managerId,
        managerIds,
        developerIds,
        verifierIds,
        dateGiven,
        lifecycleStatus,
        clientRequestId,
      } = req.body;
      const authorId = req.user?.userId;
      const missingFields = [
        !title && "title",
        !description && "description",
        !moduleId && "moduleId",
        (!managerId && (!managerIds || managerIds.length === 0)) && "managerId",
        lifecycleStatus === undefined && "lifecycleStatus",
        !authorId && "authorId",
      ].filter(Boolean);

      if (missingFields.length > 0) {
        return res.status(400).json({ error: `Missing required fields: ${missingFields.join(", ")}` });
      }

      const task = await TaskService.createTask({
        title: String(title).trim(),
        description: String(description).trim(),
        authorId: authorId as string,
        teamId,
        moduleId,
        clientId,
        managerId,
        managerIds,
        developerIds,
        verifierIds,
        dateGiven,
        lifecycleStatus: Number(lifecycleStatus),
        clientRequestId: clientRequestId !== undefined ? Number(clientRequestId) : undefined,
        plannedStartDate,
        plannedEndDate,
      });
      return res.status(201).json(task);
    } catch (error: any) {
      console.error("[TaskController.createTask]", error);
      if (
        error.message?.startsWith("Invalid") ||
        error.message?.startsWith("Missing") ||
        error.message?.includes("not found") ||
        error.message?.includes("selected module")
      ) {
        return res.status(400).json({ error: error.message });
      }
      if (error.message?.startsWith("Only")) return res.status(403).json({ error: error.message });
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async getTasks(req: AuthenticatedRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).json({ error: "Unauthorized" });
      const includeDeleted = req.query.includeDeleted === "true";
      const tasks = await TaskService.getTasks(req.user.role, req.user.userId, includeDeleted);
      return res.status(200).json(tasks);
    } catch (error) {
      console.error(error);
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async getTaskById(req: AuthenticatedRequest, res: Response) {
    try {
      const id = req.params.id as string;
      const task = await TaskService.getTaskById(id);
      if (!task) return res.status(404).json({ error: "Task not found" });

      if (!req.user) return res.status(401).json({ error: "Unauthorized" });
      try {
        await TaskService.checkReadPermission(task, req.user.userId, req.user.role);
      } catch (err: any) {
        return res.status(403).json({ error: err.message });
      }

      return res.status(200).json(task);
    } catch (error) {
      console.error(error);
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async updateStatus(req: AuthenticatedRequest, res: Response) {
    try {
      const { status, reason } = req.body;
      const actorId = req.user?.userId;
      const taskId = req.params.id as string;
      if (!status || !actorId) return res.status(400).json({ error: "Missing required fields" });
      if (!Object.values(TaskStatus).includes(status)) return res.status(400).json({ error: "Invalid status" });
      const updatedTask = await TaskService.updateStatus(taskId, actorId, status as TaskStatus, reason);
      return res.status(200).json(updatedTask);
    } catch (error: any) {
      console.error(error);
      if (error.message === "Task not found") return res.status(404).json({ error: error.message });
      if (error.message?.startsWith("Forbidden")) return res.status(403).json({ error: error.message });
      if (error.message?.startsWith("Invalid") || error.message?.startsWith("Cannot")) return res.status(400).json({ error: error.message });
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async assignTask(req: AuthenticatedRequest, res: Response) {
    try {
      const { assigneeId } = req.body;
      const actorId = req.user?.userId;
      const taskId = req.params.id as string;
      if (!assigneeId || !actorId) return res.status(400).json({ error: "Missing required fields" });
      const updatedTask = await TaskService.assignTask(taskId, assigneeId, actorId);
      return res.status(200).json(updatedTask);
    } catch (error: any) {
      console.error(error);
      if (error.message === "Task not found") return res.status(404).json({ error: error.message });
      if (error.message?.startsWith("Forbidden")) return res.status(403).json({ error: error.message });
      if (error.message?.startsWith("Invalid") || error.message?.includes("module")) return res.status(400).json({ error: error.message });
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async softDeleteTask(req: AuthenticatedRequest, res: Response) {
    try {
      const actorId = req.user?.userId;
      const taskId = req.params.id as string;
      if (!actorId) return res.status(400).json({ error: "Missing required fields" });
      const task = await TaskService.softDeleteTask(taskId, actorId);
      return res.status(200).json(task);
    } catch (error: any) {
      console.error(error);
      if (error.message === "Task not found") return res.status(404).json({ error: error.message });
      if (error.message?.startsWith("Only")) return res.status(403).json({ error: error.message });
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async restoreTask(req: AuthenticatedRequest, res: Response) {
    try {
      const actorId = req.user?.userId;
      const taskId = req.params.id as string;
      if (!actorId) return res.status(400).json({ error: "Missing required fields" });
      const task = await TaskService.restoreTask(taskId, actorId);
      return res.status(200).json(task);
    } catch (error: any) {
      console.error(error);
      if (error.message === "Task not found") return res.status(404).json({ error: error.message });
      if (error.message?.startsWith("Only")) return res.status(403).json({ error: error.message });
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async addComment(req: AuthenticatedRequest, res: Response) {
    try {
      const actorId = req.user?.userId;
      const taskId = req.params.id as string;
      const { content, files } = req.body;
      if (!actorId || !content) return res.status(400).json({ error: "Missing required fields" });
      const task = await TaskService.addComment(taskId, actorId, String(content), files);
      return res.status(201).json(task);
    } catch (error: any) {
      console.error(error);
      if (error.message === "Task not found") return res.status(404).json({ error: error.message });
      if (error.message?.startsWith("Cannot") || error.message?.includes("required")) return res.status(400).json({ error: error.message });
      return res.status(500).json({ error: "Internal server error" });
    }
  }

  static async updateTaskDetails(req: AuthenticatedRequest, res: Response) {
    try {
      const actorId = req.user?.userId;
      const taskId = req.params.id as string;
      const {
        title,
        description,
        plannedStartDate,
        plannedEndDate,
        clientId,
        clientRequestId,
        managerId,
        managerIds,
        developers,
        developerIds,
        verifiers,
        verifierIds,
        moduleId,
        dateGiven,
        dateStarted,
        dateEnded,
        status,
        reason
      } = req.body;
      if (!actorId) return res.status(400).json({ error: "Missing required fields" });
      const updated = await TaskService.updateTaskDetails(taskId, {
        title,
        description,
        plannedStartDate,
        plannedEndDate,
        clientId,
        clientRequestId: clientRequestId !== undefined ? Number(clientRequestId) : undefined,
        managerId,
        managerIds,
        developers,
        developerIds,
        verifiers,
        verifierIds,
        moduleId,
        dateGiven,
        dateStarted,
        dateEnded,
        status,
        reason
      }, actorId);
      return res.status(200).json(updated);
    } catch (error: any) {
      console.error(error);
      if (error.message?.startsWith("Forbidden")) {
        return res.status(403).json({ error: error.message });
      }
      if (
        error.message?.startsWith("Invalid") ||
        error.message?.startsWith("Missing") ||
        error.message?.includes("must belong") ||
        error.message?.includes("not found")
      ) {
        return res.status(400).json({ error: error.message });
      }
      return res.status(500).json({ error: "Internal server error" });
    }
  }
}

File: backend-spring/src/main/java/com/patchflow/controller/TaskController.java
package com.patchflow.controller;

import com.patchflow.config.Auth;
import com.patchflow.entity.User;
import com.patchflow.service.TaskService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/tasks")
@RequiredArgsConstructor
public class TaskController {

    private final TaskService taskService;

    @PostMapping
    public ResponseEntity<?> createTask(@RequestBody Map<String, Object> body, HttpServletRequest req) {
        User user = Auth.requireRole(req, "SUPER_ADMIN","ADMIN","CLIENT","MANAGER","DEVELOPER");
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> result = taskService.createTask(
                    user.getUserId(),
                    (String) body.get("title"),
                    (String) body.get("description"),
                    (String) body.get("moduleId"),
                    (String) body.get("teamId"),
                    (String) body.get("clientId"),
                    body.get("clientRequestId") != null ? ((Number) body.get("clientRequestId")).intValue() : null,
                    body.get("managerIds") != null ? (List<String>) body.get("managerIds") :
                            (body.get("managerId") != null ? List.of((String) body.get("managerId")) : null),
                    (List<String>) body.get("developerIds"),
                    (List<String>) body.get("verifierIds"),
                    (String) body.get("dateGiven"),
                    body.get("lifecycleStatus") != null ? ((Number) body.get("lifecycleStatus")).intValue() : 0,
                    (String) body.get("plannedStartDate"),
                    (String) body.get("plannedEndDate")
            );
            return ResponseEntity.status(HttpStatus.CREATED).body(result);
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }

    @GetMapping
    public ResponseEntity<?> getTasks(@RequestParam(defaultValue = "false") boolean includeDeleted, HttpServletRequest req) {
        User user = Auth.require(req);
        return ResponseEntity.ok(taskService.getTasks(user.getRole(), user.getUserId(), includeDeleted));
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getTaskById(@PathVariable String id, HttpServletRequest req) {
        User user = Auth.require(req);
        Map<String, Object> task = taskService.getTaskById(id);
        if (task == null) return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Task not found"));
        try {
            taskService.checkReadPermission(task, user.getUserId(), user.getRole());
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
        return ResponseEntity.ok(task);
    }

    @PatchMapping("/{id}/status")
    public ResponseEntity<?> updateStatus(@PathVariable String id, @RequestBody Map<String, Object> body, HttpServletRequest req) {
        User user = Auth.require(req);
        String status = (String) body.get("status");
        String reason = (String) body.get("reason");
        if (status == null) return ResponseEntity.badRequest().body(Map.of("error", "Missing required fields"));
        try {
            return ResponseEntity.ok(taskService.updateStatus(id, user.getUserId(), status, reason));
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }

    @PostMapping("/{id}/comments")
    public ResponseEntity<?> addComment(@PathVariable String id, @RequestBody Map<String, Object> body, HttpServletRequest req) {
        User user = Auth.require(req);
        String content = (String) body.get("content");
        if (content == null) return ResponseEntity.badRequest().body(Map.of("error", "Missing required fields"));
        try {
            return ResponseEntity.status(HttpStatus.CREATED).body(taskService.addComment(id, user.getUserId(), content, body.get("files")));
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> softDelete(@PathVariable String id, HttpServletRequest req) {
        User user = Auth.require(req);
        try {
            return ResponseEntity.ok(taskService.softDeleteTask(id, user.getUserId()));
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }

    @PostMapping("/{id}/restore")
    public ResponseEntity<?> restore(@PathVariable String id, HttpServletRequest req) {
        User user = Auth.require(req);
        try {
            return ResponseEntity.ok(taskService.restoreTask(id, user.getUserId()));
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }

    @PostMapping("/{id}/assign")
    public ResponseEntity<?> assign(@PathVariable String id, @RequestBody Map<String, String> body, HttpServletRequest req) {
        User user = Auth.require(req);
        String assigneeId = body.get("assigneeId");
        if (assigneeId == null) return ResponseEntity.badRequest().body(Map.of("error", "Missing required fields"));
        try {
            return ResponseEntity.ok(taskService.assignTask(id, assigneeId, user.getUserId()));
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }

    @PatchMapping("/{id}/details")
    public ResponseEntity<?> updateDetails(@PathVariable String id, @RequestBody Map<String, Object> body, HttpServletRequest req) {
        User user = Auth.require(req);
        try {
            return ResponseEntity.ok(taskService.updateTaskDetails(id, body, user.getUserId()));
        } catch (ResponseStatusException ex) {
            return ResponseEntity.status(ex.getStatusCode()).body(Map.of("error", ex.getReason()));
        }
    }
}
